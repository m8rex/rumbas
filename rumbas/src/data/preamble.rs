use crate::data::file_reference::FileString;
use crate::data::optional_overwrite::{Noneable, OptionalOverwrite};
use crate::data::to_numbas::{NumbasResult, ToNumbas};
use serde::{Deserialize, Serialize};

// TODO: maybe translatable? So different text's can be set for different languages?
optional_overwrite! {
    Preamble,
    js: FileString,
    css: FileString
}

impl ToNumbas for Preamble {
    type NumbasType = numbas::exam::Preamble;
    fn to_numbas(&self, _locale: &String) -> NumbasResult<numbas::exam::Preamble> {
        let empty_fields = self.empty_fields();
        if empty_fields.is_empty() {
            Ok(numbas::exam::Preamble::new(
                self.js.clone().unwrap().get_content(),
                self.css.clone().unwrap().get_content(),
            ))
        } else {
            Err(empty_fields)
        }
    }
}
