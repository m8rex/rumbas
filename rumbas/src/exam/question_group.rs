use crate::question::Question;
use crate::support::optional_overwrite::*;
use crate::support::template::{Value, ValueType};
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use crate::support::translatable::TranslatableString;
use crate::support::yaml::YamlError;
use sanitize_filename::sanitize;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

optional_overwrite! {
    pub struct QuestionGroup {
        /// The name
        name: TranslatableString,
        /// The strategy to use to pick the questions to show
        #[serde(flatten)]
        picking_strategy: PickingStrategy,
        /// The questions
        questions: Vec<Value<QuestionPath>>
    }
}

impl ToNumbas<numbas::exam::question_group::QuestionGroup> for QuestionGroup {
    fn to_numbas(&self, locale: &str) -> numbas::exam::question_group::QuestionGroup {
        numbas::exam::question_group::QuestionGroup {
            name: Some(self.name.to_numbas(locale)),
            picking_strategy: self.picking_strategy.to_numbas(locale),
            questions: self.questions.to_numbas(locale),
        }
    }
}

impl ToRumbas<QuestionGroup> for numbas::exam::question_group::QuestionGroup {
    fn to_rumbas(&self) -> QuestionGroup {
        QuestionGroup {
            name: self.name.clone().unwrap_or_default().to_rumbas(),
            picking_strategy: self.picking_strategy.to_rumbas(),
            questions: self.questions.to_rumbas(),
        }
    }
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone, PartialEq)]
#[serde(tag = "picking_strategy")]
pub enum PickingStrategy {
    #[serde(rename = "all_ordered")]
    AllOrdered,
    #[serde(rename = "all_shuffled")]
    AllShuffled,
    #[serde(rename = "random_subset")]
    RandomSubset { pick_questions: usize },
}
impl_optional_overwrite!(PickingStrategy);

impl ToNumbas<numbas::exam::question_group::QuestionGroupPickingStrategy> for PickingStrategy {
    fn to_numbas(
        &self,
        _locale: &str,
    ) -> numbas::exam::question_group::QuestionGroupPickingStrategy {
        match self {
            PickingStrategy::AllOrdered => {
                numbas::exam::question_group::QuestionGroupPickingStrategy::AllOrdered
            }
            PickingStrategy::AllShuffled => {
                numbas::exam::question_group::QuestionGroupPickingStrategy::AllShuffled
            }
            PickingStrategy::RandomSubset { pick_questions } => {
                numbas::exam::question_group::QuestionGroupPickingStrategy::RandomSubset {
                    pick_questions: *pick_questions,
                }
            }
        }
    }
}

impl ToRumbas<PickingStrategy> for numbas::exam::question_group::QuestionGroupPickingStrategy {
    fn to_rumbas(&self) -> PickingStrategy {
        match self {
            numbas::exam::question_group::QuestionGroupPickingStrategy::AllOrdered => {
                PickingStrategy::AllOrdered
            }
            numbas::exam::question_group::QuestionGroupPickingStrategy::AllShuffled => {
                PickingStrategy::AllShuffled
            }
            numbas::exam::question_group::QuestionGroupPickingStrategy::RandomSubset {
                pick_questions,
            } => PickingStrategy::RandomSubset {
                pick_questions: *pick_questions,
            },
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(try_from = "String")]
#[serde(into = "String")]
pub struct QuestionPath {
    pub question_name: String,
    pub question_data: Question,
}

impl RumbasCheck for QuestionPath {
    fn check(&self, locale: &str) -> RumbasCheckResult {
        self.question_data.check(locale)
    }
}

impl OptionalOverwrite<QuestionPath> for QuestionPath {
    fn overwrite(&mut self, _other: &Self) {}
    fn insert_template_value(&mut self, key: &str, val: &serde_yaml::Value) {
        self.question_data.insert_template_value(key, val);
    }
}
impl_optional_overwrite_value!(QuestionPath);

impl ToNumbas<numbas::question::question::Question> for QuestionPath {
    fn to_numbas(&self, locale: &str) -> numbas::question::question::Question {
        self.question_data
            .clone()
            .to_numbas_with_name(locale, self.question_name.clone())
    }
}

impl ToRumbas<QuestionPath> for numbas::question::question::Question {
    fn to_rumbas(&self) -> QuestionPath {
        QuestionPath {
            question_name: sanitize(&self.name),
            question_data: self.to_rumbas(),
        }
    }
}

impl JsonSchema for QuestionPath {
    fn schema_name() -> String {
        "QuestionPath".to_owned()
    }

    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
        gen.subschema_for::<String>()
    }
}

impl std::convert::TryFrom<String> for QuestionPath {
    type Error = YamlError;

    fn try_from(s: String) -> Result<Self, Self::Error> {
        let question_data = Question::from_name(&s).map_err(|e| e)?;
        Ok(QuestionPath {
            question_name: s,
            question_data,
        })
    }
}

impl std::convert::From<QuestionPath> for String {
    fn from(q: QuestionPath) -> Self {
        q.question_name
    }
}
