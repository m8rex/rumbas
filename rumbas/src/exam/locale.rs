use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::Deserialize;
use serde::Serialize;
use comparable::Comparable;

#[derive(Input, Overwrite, RumbasCheck, Examples)]
#[input(name = "LocaleInput")]
#[derive(Serialize, Deserialize, Comparable, Debug, Clone, JsonSchema, PartialEq)]
pub struct Locale {
    pub name: String, //TODO: document names best used for shareability?
    /// The locale to use in the Numbas interface
    pub numbas_locale: SupportedLocale,
}

macro_rules! create_support_locale {
    ($($name: ident => $key: literal),*) => {
        #[derive(Input, Overwrite, RumbasCheck, Examples)]
        #[input(name = "SupportedLocaleInput")]
        /// Locales supported by Numbas
        #[derive(Serialize, Deserialize, Comparable, JsonSchema, Debug, Clone, Copy, PartialEq)]
        pub enum SupportedLocale {
            $(
                #[serde(rename = $key)]
                $name
            ),*
        }

        //TODO? macro to reduce duplication?
        impl SupportedLocale {
            pub fn to_str(self) -> &'static str {
                match self {
                    $(SupportedLocale::$name => $key),*
                }
            }
        }

    }
}

create_support_locale! {
    ArSA => "ar-SA",
    DeDE => "de-DE",
    EnGB => "en-GB",
    EsES => "es-ES",
    FrFR => "fr-FR",
    HeIL => "he-IL",
    ItIT => "it-IT",
    JaJP => "ja-JP",
    KoKR => "ko-KR",
    NbNO => "nb-NO",
    NlNL => "nl-NL",
    PlPL => "pl-PL",
    PtBR => "pt-BR",
    SqAL => "sq-AL",
    SvSE => "sv-SE",
    TrTR => "tr-TR",
    ViVN => "vi-VN",
    ZhCN => "zg-CN"
}
