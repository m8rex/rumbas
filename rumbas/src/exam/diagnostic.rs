use crate::exam::feedback::Feedback;
use crate::exam::locale::SupportedLocale;
use crate::exam::locale::{Locale, Locales};
use crate::exam::navigation::DiagnosticNavigation;
use crate::exam::numbas_settings::NumbasSettings;
use crate::exam::question_group::QuestionGroups;
use crate::exam::question_group::QuestionPath;
use crate::exam::timing::Timing;
use crate::question::custom_part_type::CustomPartTypeDefinitionPath;
use crate::question::extension::Extensions;
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use crate::support::translatable::JMENotesTranslatableString;
use crate::support::translatable::TranslatableString;
use crate::support::translatable::TranslatableStrings;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "DiagnosticExamInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
/// A Diagnostic Exam
pub struct DiagnosticExam {
    /// All locales for which the exam should be generated
    pub locales: Locales,
    /// The name of the exam
    pub name: TranslatableString,
    /// The navigation settings for this exam
    pub navigation: DiagnosticNavigation,
    /// The timing settings for this exam
    pub timing: Timing,
    /// The feedback settings for this exam
    pub feedback: Feedback,
    /// The questions groups for this exam
    pub question_groups: QuestionGroups,
    /// The settings to set for numbas
    pub numbas_settings: NumbasSettings,
    /// The diagnostic data
    pub diagnostic: Diagnostic,
}

impl ToNumbas<numbas::exam::exam::Exam> for DiagnosticExam {
    fn to_numbas(&self, locale: &str) -> numbas::exam::exam::Exam {
        let basic_settings = self.to_numbas(locale);

        let navigation = self.navigation.to_numbas(locale);

        let timing = self.timing.to_numbas(locale);

        let feedback = self.feedback.to_numbas(locale);

        //TODO
        let functions = Some(HashMap::new());

        //TODO
        let variables = Some(HashMap::new());

        let question_groups: Vec<numbas::exam::question_group::QuestionGroup> =
            self.question_groups.to_numbas(locale);

        let resources: Vec<numbas::question::resource::Resource> = self
            .question_groups
            .iter()
            .flat_map(|qg| {
                qg.clone()
                    .questions
                    .into_iter()
                    .flat_map(|q| q.question_data.resources)
            })
            .collect::<std::collections::HashSet<_>>()
            .into_iter()
            .collect::<Vec<_>>()
            .to_numbas(locale); // TODO: extract?

        let extensions: Vec<String> = self
            .question_groups
            .iter()
            .flat_map(|qg| {
                qg.clone()
                    .questions
                    .into_iter()
                    .map(|q| q.question_data.extensions) // todo: extract?
            })
            .fold(Extensions::default(), Extensions::combine)
            .to_paths();

        let custom_part_types: Vec<numbas::question::custom_part_type::CustomPartType> = self
            .question_groups
            .iter()
            .flat_map(|qg| {
                qg.clone()
                    .questions
                    .into_iter()
                    .flat_map(|q| q.question_data.custom_part_types)
            })
            .collect::<std::collections::HashSet<_>>()
            .into_iter()
            .collect::<Vec<_>>()
            .to_numbas(locale); // todo extract?

        let diagnostic = Some(self.diagnostic.to_numbas(locale));

        numbas::exam::exam::Exam {
            basic_settings,
            resources,
            extensions,
            custom_part_types,
            navigation,
            timing,
            feedback,
            functions,
            variables,
            question_groups,
            diagnostic,
        }
    }
}

impl ToNumbas<numbas::exam::exam::BasicExamSettings> for DiagnosticExam {
    fn to_numbas(&self, locale: &str) -> numbas::exam::exam::BasicExamSettings {
        numbas::exam::exam::BasicExamSettings {
            name: self.name.to_numbas(locale),
            duration_in_seconds: self.timing.duration_in_seconds.to_numbas(locale),
            percentage_needed_to_pass: self.feedback.percentage_needed_to_pass.to_numbas(locale),
            show_question_group_names: Some(
                self.navigation.shared_data.show_names_of_question_groups,
            ),
            show_student_name: Some(self.feedback.clone().show_name_of_student),
            allow_printing: Some(self.navigation.shared_data.allow_printing),
        }
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "DiagnosticInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
/// Information needed for a diagnostic test
pub struct Diagnostic {
    /// The script to use
    script: DiagnosticScript,
    /// The learning objectives,
    objectives: LearningObjectives,
    /// The learning topics
    topics: LearningTopics,
}

impl ToNumbas<numbas::exam::diagnostic::Diagnostic> for Diagnostic {
    fn to_numbas(&self, locale: &str) -> numbas::exam::diagnostic::Diagnostic {
        numbas::exam::diagnostic::Diagnostic {
            knowledge_graph: self.to_numbas(locale),
            script: self.script.to_numbas(locale),
            custom_script: self.script.to_numbas(locale),
        }
    }
}

impl ToNumbas<numbas::exam::diagnostic::DiagnosticKnowledgeGraph> for Diagnostic {
    fn to_numbas(&self, locale: &str) -> numbas::exam::diagnostic::DiagnosticKnowledgeGraph {
        numbas::exam::diagnostic::DiagnosticKnowledgeGraph {
            topics: self.topics.to_numbas(locale),
            learning_objectives: self.objectives.to_numbas(locale),
        }
    }
}

impl ToRumbas<Diagnostic> for numbas::exam::diagnostic::Diagnostic {
    fn to_rumbas(&self) -> Diagnostic {
        Diagnostic {
            script: self.to_rumbas(),
            objectives: self.knowledge_graph.learning_objectives.to_rumbas(),
            topics: self.knowledge_graph.topics.to_rumbas(),
        }
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "DiagnosticScriptInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum DiagnosticScript {
    Mastery,
    Diagnosys,
    Custom(JMENotesTranslatableString),
}

impl ToNumbas<numbas::exam::diagnostic::DiagnosticScript> for DiagnosticScript {
    fn to_numbas(&self, _locale: &str) -> numbas::exam::diagnostic::DiagnosticScript {
        match self {
            DiagnosticScript::Mastery => numbas::exam::diagnostic::DiagnosticScript::Mastery,
            DiagnosticScript::Custom(_) => numbas::exam::diagnostic::DiagnosticScript::Custom,
            DiagnosticScript::Diagnosys => numbas::exam::diagnostic::DiagnosticScript::Diagnosys,
        }
    }
}

impl ToNumbas<numbas::jme::JMENotesString> for DiagnosticScript {
    fn to_numbas(&self, locale: &str) -> numbas::jme::JMENotesString {
        match self {
            DiagnosticScript::Custom(s) => s.to_numbas(locale),
            DiagnosticScript::Diagnosys => Default::default(),
            DiagnosticScript::Mastery => Default::default(),
        }
    }
}

impl ToRumbas<DiagnosticScript> for numbas::exam::diagnostic::Diagnostic {
    fn to_rumbas(&self) -> DiagnosticScript {
        match self.script {
            numbas::exam::diagnostic::DiagnosticScript::Mastery => DiagnosticScript::Mastery,
            numbas::exam::diagnostic::DiagnosticScript::Diagnosys => DiagnosticScript::Diagnosys,
            numbas::exam::diagnostic::DiagnosticScript::Custom => {
                DiagnosticScript::Custom(self.custom_script.clone().into())
            }
        }
    }
}

pub type LearningObjectivesInput = Vec<Value<LearningObjectiveInput>>;
pub type LearningObjectives = Vec<LearningObjective>;

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "LearningObjectiveInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq)]
/// A Learning Objective
pub struct LearningObjective {
    /// The name
    name: TranslatableString,
    /// A description
    description: TranslatableString,
}

impl ToNumbas<numbas::exam::diagnostic::DiagnosticKnowledgeGraphLearningObjective>
    for LearningObjective
{
    fn to_numbas(
        &self,
        locale: &str,
    ) -> numbas::exam::diagnostic::DiagnosticKnowledgeGraphLearningObjective {
        numbas::exam::diagnostic::DiagnosticKnowledgeGraphLearningObjective {
            name: self.name.to_numbas(locale),
            description: self.description.to_numbas(locale),
        }
    }
}

impl ToRumbas<LearningObjective>
    for numbas::exam::diagnostic::DiagnosticKnowledgeGraphLearningObjective
{
    fn to_rumbas(&self) -> LearningObjective {
        LearningObjective {
            name: self.name.to_rumbas(),
            description: self.description.to_rumbas(),
        }
    }
}

pub type LearningTopicsInput = Vec<Value<LearningTopicInput>>;
pub type LearningTopics = Vec<LearningTopic>;

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "LearningTopicInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq)]
/// A learning Topic
pub struct LearningTopic {
    /// The name
    name: TranslatableString,
    /// A description
    description: TranslatableString,
    /// List of names of objectives
    objectives: TranslatableStrings,
    /// List of names of topic on which this topic depends
    depends_on: TranslatableStrings,
}

impl ToNumbas<numbas::exam::diagnostic::DiagnosticKnowledgeGraphTopic> for LearningTopic {
    fn to_numbas(&self, locale: &str) -> numbas::exam::diagnostic::DiagnosticKnowledgeGraphTopic {
        numbas::exam::diagnostic::DiagnosticKnowledgeGraphTopic {
            name: self.name.to_numbas(locale),
            description: self.description.to_numbas(locale),
            learning_objectives: self.objectives.to_numbas(locale),
            depends_on: self.depends_on.to_numbas(locale),
        }
    }
}

impl ToRumbas<LearningTopic> for numbas::exam::diagnostic::DiagnosticKnowledgeGraphTopic {
    fn to_rumbas(&self) -> LearningTopic {
        LearningTopic {
            name: self.name.to_rumbas(),
            description: self.description.to_rumbas(),
            objectives: self.learning_objectives.to_rumbas(),
            depends_on: self.depends_on.to_rumbas(),
        }
    }
}

/// Converts a diagnostic numbas exam to a DiagnosticExam and extracts questions and
/// custom_part_types
pub fn convert_diagnostic_numbas_exam(
    exam: numbas::exam::exam::Exam,
) -> (
    DiagnosticExam,
    Vec<QuestionPath>,
    Vec<CustomPartTypeDefinitionPath>,
) {
    let question_groups: Vec<_> = exam.question_groups.to_rumbas();
    let custom_part_types = exam.custom_part_types.to_rumbas();
    (
        DiagnosticExam {
            locales: vec![Locale {
                name: "en".to_string(),
                numbas_locale: SupportedLocale::EnGB,
            }], // todo: argument?
            name: exam.basic_settings.name.to_rumbas(),
            navigation: exam.to_rumbas(),
            timing: exam.to_rumbas(),
            feedback: exam.to_rumbas(),
            question_groups: question_groups.clone(),
            numbas_settings: NumbasSettings {
                theme: "default".to_string(),
            }, // todo: argument?
            diagnostic: exam.diagnostic.unwrap().to_rumbas(), // Always set for a diagnostic exam
        },
        question_groups
            .into_iter()
            .flat_map(|qg| qg.questions)
            .collect(),
        custom_part_types,
    )
}
