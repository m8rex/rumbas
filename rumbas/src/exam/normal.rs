use crate::exam::feedback::Feedback;
use crate::exam::locale::Locale;
use crate::exam::locale::SupportedLocale;
use crate::exam::navigation::NormalNavigation;
use crate::exam::numbas_settings::NumbasSettings;
use crate::exam::question_group::QuestionGroup;
use crate::exam::question_group::QuestionPath;
use crate::exam::timing::Timing;
use crate::question::custom_part_type::CustomPartTypeDefinitionPath;
use crate::question::extension::Extensions;
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use crate::support::translatable::TranslatableString;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_diff::{Apply, Diff, SerdeDiff};
use std::collections::HashMap;

// TODO: remove duplication of NormalExam & Diagnostic Exam?
#[derive(Input, Overwrite, RumbasCheck, Examples)]
#[input(name = "NormalExamInput")]
#[derive(Serialize, Deserialize, SerdeDiff, Debug, Clone, JsonSchema, PartialEq)]
/// An Exam
pub struct NormalExam {
    /// All locales for which the exam should be generated
    pub locales: Vec<Locale>,
    /// The name of the exam
    pub name: TranslatableString,
    /// The navigation settings for this exam
    pub navigation: NormalNavigation,
    /// The timing settings for this exam
    pub timing: Timing,
    /// The feedback settings for this exam
    pub feedback: Feedback,
    /// The questions groups for this exam
    pub question_groups: Vec<QuestionGroup>,
    /// The settings to set for numbas
    pub numbas_settings: NumbasSettings,
}

impl ToNumbas<numbas::exam::Exam> for NormalExam {
    fn to_numbas(&self, locale: &str) -> numbas::exam::Exam {
        let basic_settings = self.to_numbas(locale);
        let navigation = self.navigation.to_numbas(locale);

        let timing = self.timing.to_numbas(locale);

        let feedback = self.feedback.to_numbas(locale);

        //TODO
        let functions = Some(HashMap::new());

        //TODO
        let variables = Some(HashMap::new());

        let question_groups: Vec<numbas::exam::question_group::QuestionGroup> =
            self.question_groups.to_numbas(locale);

        let resources: Vec<numbas::question::resource::Resource> = self
            .question_groups
            .iter()
            .flat_map(|qg| {
                qg.clone()
                    .questions
                    .into_iter()
                    .flat_map(|q| q.data.resources)
            })
            .collect::<std::collections::HashSet<_>>()
            .into_iter()
            .collect::<Vec<_>>()
            .to_numbas(locale);

        let extensions: Vec<String> = self
            .question_groups
            .iter()
            .flat_map(|qg| qg.clone().questions.into_iter().map(|q| q.data.extensions))
            .fold(Extensions::default(), Extensions::combine)
            .to_paths();

        let custom_part_types: Vec<numbas::question::custom_part_type::CustomPartType> = self
            .question_groups
            .clone()
            .iter()
            .flat_map(|qg| {
                qg.clone()
                    .questions
                    .into_iter()
                    .flat_map(|q| q.data.custom_part_types)
            })
            .collect::<std::collections::HashSet<_>>()
            .into_iter()
            .collect::<Vec<_>>()
            .to_numbas(locale);

        numbas::exam::Exam {
            basic_settings,
            resources,
            extensions,
            custom_part_types,
            navigation,
            timing,
            feedback,
            functions,
            variables,
            question_groups,
            diagnostic: None,
        }
    }
}

impl ToNumbas<numbas::exam::BasicExamSettings> for NormalExam {
    fn to_numbas(&self, locale: &str) -> numbas::exam::BasicExamSettings {
        numbas::exam::BasicExamSettings {
            name: self.name.to_numbas(locale),
            duration_in_seconds: self.timing.duration_in_seconds.to_numbas(locale),
            percentage_needed_to_pass: self.feedback.percentage_needed_to_pass.to_numbas(locale),
            show_question_group_names: Some(
                self.navigation
                    .to_shared_data()
                    .show_names_of_question_groups,
            ),
            show_student_name: Some(self.feedback.show_name_of_student),
            allow_printing: Some(self.navigation.to_shared_data().allow_printing),
        }
    }
}

/// Converts a normal numbas exam to a NormalExam
pub fn convert_normal_numbas_exam(
    exam: numbas::exam::Exam,
) -> (
    NormalExam,
    Vec<QuestionPath>,
    Vec<CustomPartTypeDefinitionPath>,
) {
    let question_groups: Vec<_> = exam.question_groups.to_rumbas();
    let custom_part_types = exam.custom_part_types.to_rumbas();
    (
        NormalExam {
            locales: vec![Locale {
                name: "en".to_string(),
                numbas_locale: SupportedLocale::EnGB,
            }], // todo: argument?
            name: exam.basic_settings.name.to_rumbas(),
            navigation: exam.to_rumbas(),
            timing: exam.to_rumbas(),
            feedback: exam.to_rumbas(),
            question_groups: question_groups.clone(),
            numbas_settings: NumbasSettings {
                theme: "default".to_string(),
            }, // todo: argument?
        },
        question_groups
            .into_iter()
            .flat_map(|qg| qg.questions.into_iter())
            .collect(),
        custom_part_types,
    )
}
