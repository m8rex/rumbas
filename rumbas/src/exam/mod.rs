//! Contains all the exam types

pub mod diagnostic;
pub mod feedback;
pub mod locale;
pub mod navigation;
pub mod normal;
pub mod numbas_settings;
pub mod question_group;
pub mod timing;

use crate::exam::diagnostic::convert_diagnostic_numbas_exam;
use crate::exam::diagnostic::DiagnosticExam;
use crate::exam::locale::Locale;
use crate::exam::normal::convert_normal_numbas_exam;
use crate::exam::normal::NormalExam;
use crate::exam::question_group::QuestionPath;
use crate::question::custom_part_type::CustomPartTypeDefinitionPath;
use crate::support::template::{TemplateFile, TemplateFileInputEnum};
use crate::support::to_numbas::ToNumbas;
use crate::support::translatable::TranslatableString;
use crate::support::yaml::YamlError;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fmt::Display;
use std::fs;
use std::path::Path;

#[derive(Input, Overwrite, RumbasCheck, Examples)]
#[input(name = "ExamInput")]
#[input(test)]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type")]
pub enum Exam {
    Normal(NormalExam),
    Diagnostic(DiagnosticExam),
}

impl ToNumbas<numbas::exam::Exam> for Exam {
    fn to_numbas(&self, locale: &str) -> numbas::exam::Exam {
        match self {
            Exam::Normal(n) => n.to_numbas(locale),
            Exam::Diagnostic(n) => n.to_numbas(locale),
        }
    }
}

impl Exam {
    pub fn locales(&self) -> Vec<Locale> {
        match self {
            Exam::Normal(n) => n.locales.clone(),
            Exam::Diagnostic(n) => n.locales.clone(),
        }
    }

    pub fn numbas_settings(&self) -> crate::exam::numbas_settings::NumbasSettings {
        match self {
            Exam::Normal(n) => n.numbas_settings.clone(),
            Exam::Diagnostic(n) => n.numbas_settings.clone(),
        }
    }
}
impl ExamInput {
    pub fn from_file(file: &Path) -> Result<ExamInput, ParseError> {
        use ExamFileTypeInput::*;
        let input: std::result::Result<ExamFileTypeInput, serde_yaml::Error> =
            if file.starts_with(crate::EXAMS_FOLDER) {
                let yaml = fs::read_to_string(file).map_err(ParseError::IOError)?;
                serde_yaml::from_str(&yaml)
            } else if file.starts_with(crate::QUESTIONS_FOLDER) {
                let mut data = HashMap::new();
                data.insert(
                    "question".to_string(),
                    serde_yaml::Value::String(
                        file.with_extension("")
                            .strip_prefix(crate::QUESTIONS_FOLDER)
                            .unwrap()
                            .to_string_lossy()
                            .into_owned(),
                    )
                    .into(),
                );
                let t = TemplateFile {
                    relative_template_path: crate::QUESTION_PREVIEW_TEMPLATE_NAME.to_string(),
                    data,
                };
                Ok(Template(TemplateFileInputEnum::from_normal(t)))
            } else {
                panic!(
                    "{} should start with {}/ or {}/",
                    file.display(),
                    crate::EXAMS_FOLDER,
                    crate::QUESTIONS_FOLDER
                );
            };
        input
            .map(|e| match e {
                Normal(e) => Ok(ExamInput::Normal(e)),
                Diagnostic(e) => Ok(ExamInput::Diagnostic(e)),
                Template(t_val) => {
                    let t = t_val.to_normal();
                    let template_file = Path::new(crate::EXAM_TEMPLATES_FOLDER)
                        .join(format!("{}.yaml", t.relative_template_path)); // TODO: check for missing fields.....
                    let template_yaml = fs::read_to_string(&template_file).expect(
                        &format!(
                            "Failed to read {}",
                            template_file.to_str().map_or("invalid filename", |s| s)
                        )[..],
                    );

                    let mut exam: ExamInput = serde_yaml::from_str(&template_yaml).unwrap();
                    t.data.iter().for_each(|(k, v)| {
                        exam.insert_template_value(k, &v.0);
                    });
                    Ok(exam)
                }
            })
            .and_then(std::convert::identity) //flatten result is currently only possible in nightly
            .map_err(|e| ParseError::YamlError(YamlError::from(e, file.to_path_buf())))
    }
}

#[derive(Debug, Display)]
pub enum ParseError {
    YamlError(YamlError),
    IOError(std::io::Error),
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "ExamFileTypeInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type")]
pub enum ExamFileType {
    Template(TemplateFile),
    Normal(NormalExam),
    Diagnostic(DiagnosticExam),
}

impl ExamFileTypeInput {
    pub fn to_yaml(&self) -> serde_yaml::Result<String> {
        serde_yaml::to_string(self)
    }
}

impl ExamFileType {
    pub fn to_yaml(&self) -> serde_yaml::Result<String> {
        ExamFileTypeInput::from_normal(self.to_owned()).to_yaml()
    }
}

/// Convert a numbas exam to rumbas data
/// Returns the name of the exam, the resulting exam (as ExamFileType)
/// and vectors of questions and custom part type definitions
pub fn convert_numbas_exam(
    exam: numbas::exam::Exam,
) -> (
    String,
    ExamFileType,
    Vec<QuestionPath>,
    Vec<CustomPartTypeDefinitionPath>,
) {
    let (name, exam, qgs, cpts) = match exam.navigation.navigation_mode {
        numbas::exam::navigation::NavigationMode::Diagnostic(ref _d) => {
            let (exam, qgs, cpts) = convert_diagnostic_numbas_exam(exam);
            (exam.name.clone(), ExamFileType::Diagnostic(exam), qgs, cpts)
        }
        _ => {
            let (exam, qgs, cpts) = convert_normal_numbas_exam(exam);
            (exam.name.clone(), ExamFileType::Normal(exam), qgs, cpts)
        }
    };
    (
        {
            if let TranslatableString::NotTranslated(n) = name {
                n.get_content("").unwrap()
            } else {
                panic!("Should not happen");
            }
        },
        exam,
        qgs,
        cpts,
    )
}
