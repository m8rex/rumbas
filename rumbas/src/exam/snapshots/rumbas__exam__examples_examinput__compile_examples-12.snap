---
source: rumbas/src/exam/mod.rs
expression: "&example"
---
type: normal
name:
  content: "file:path/to/file"
  placeholders:
    placeholder1:
      content:
        en: en value of placeholder1
        nl: nl value of placeholder1
      placeholders: {}
    placeholder2:
      content:
        en: en version of placeholder2
        nl: nl version of placeholder2
      placeholders: {}
locales: ~
navigation:
  allow_printing: "template:template_key"
  browsing_enabled: ~
  can_move_to_previous: ~
  can_regenerate: "template:template_key"
  confirm_when_leaving: "template:template_key"
  mode: sequential
  on_leave:
    action: prevent_if_not_attempted
    message:
      content: "file:path/to/file"
      placeholders:
        placeholder1:
          content:
            en: en value of placeholder1
            nl: nl value of placeholder1
          placeholders: {}
        placeholder2:
          content:
            en: en version of placeholder2
            nl: nl version of placeholder2
          placeholders: {}
  show_names_of_question_groups: "template:template_key"
  show_results_page: never
  show_steps: "template:template_key"
  show_title_page: "template:template_key"
  start_password: none
timing: "template:template_key"
feedback:
  percentage_needed_to_pass: 1.2
  show_name_of_student: false
  show_current_marks: false
  show_maximum_marks: false
  show_answer_state: false
  allow_reveal_answer: false
  review:
    show_score: false
    show_feedback: false
    show_expected_answer: false
    show_advice: false
  advice:
    content:
      nonjmetext§: "file:path/to/file"
    placeholders:
      placeholder1:
        content:
          en: en value of placeholder1
          nl: nl value of placeholder1
        placeholders: {}
      placeholder2:
        content:
          en: en version of placeholder2
          nl: nl version of placeholder2
        placeholders: {}
  intro:
    content:
      nonjmetext§: "file:path/to/file"
    placeholders:
      placeholder1:
        content:
          en: en value of placeholder1
          nl: nl value of placeholder1
        placeholders: {}
      placeholder2:
        content:
          en: en version of placeholder2
          nl: nl version of placeholder2
        placeholders: {}
  feedback_messages:
    - message: nonjmetext§
      threshold: nonjmetext§
    - message: "template:template_key"
      threshold: "template:template_key"
    - message:
        template_key: template_key
        default_value: ~
      threshold:
        template_key: template_key
        default_value: ~
    - message: ~
      threshold: ~
    - "template:template_key"
    - template_key: template_key
      default_value: ~
question_groups: ~
numbas_settings: "template:template_key"

